uaclient/security.py

```diff
@@ -296,46 +342,36 @@ def fix_security_issue_id(cfg: UAConfig, issue_id: str) -> None:
 
 
 def get_usn_affected_packages_status(
-    client: UASecurityClient, issue_id: str
+    usn: USN, installed_packages: "Dict[str,str]"
 ) -> "Dict[str, CVEPackageStatus]":
-    """Walk CVEs related to a USN and determine if all are fixed."""
-    affected_pkg_versions = {}  # type: Dict[str, CVEPackageStatus]
-    try:
-        usn = client.get_notice(notice_id=issue_id)
-    except SecurityAPIError as e:
-        raise exceptions.UserFacingError(str(e))
-    print(
-        status.MESSAGE_SECURITY_URL.format(
-            issue=usn.id,
-            title=usn.title,
-            url_path="notices/{}.json".format(usn.id.lower()),
-        )
-    )
-    return affected_pkg_versions  # TODO(walk USN CVEs for affected packages)
+    """Walk CVEs related to a USN and return a dict of all affected packages.
+
+    :return: Dict keyed on source package name, with active CVEPackageStatus
+        for the current Ubuntu release.
+    """
+    affected_pkgs = {}  # type: Dict[str, CVEPackageStatus]
+    for cve in usn.get_cves_metadata():
+        for pkg_name, pkg_status in get_cve_affected_packages_status(
+            cve, installed_packages
+        ).items():
+            if pkg_name not in affected_pkgs:
```

Comment:
If we only have the else, we can't find the `current_ver = affected_pkgs[pkg_name].fixed_version` because affected_pkgs.get(pkg_name) is None instead of a CVEPackageStatus instance.

Reply:
Sorry @blackboxsw I think you have already explained that to me, but what is really bugging me here is that why looking at the code, `get_cve_affected_packages_status` could be returning a dict with duplicated packages on it. And this is something that I am not getting here on why that method would do that, since we only add pkgs to the dict if they are installed in the system.

Reply:
@lucasmoura, now I understand your query. So, get_cve_affected_packages_status will only return unique package names for this particular CVE that happen to affect this running ubuntu release. But, we are in a for loop which is walking through all related CVEs for a given USN. Since multiple CVEs are potentially walked, it's possible that two different CVEs may have overlapping/duplicate affected packages listed and competing minimal package versions required. If CVE1 requires apt: 1.6.12 and CVE2 requires apt: 1.6.13  we want to make sure we represent the max of those versions recommendations to ensure we are effectively patched.

Reply:
Make complete sense now. I missed the fact that cve loop might have colliding packages. Thanks for the explanation @blackboxsw